Provide EVM builtins `loadimmutable` and `setimmutable`

Marenz opened the pull request:
refs #8551

This is just a draft, tests are missing.
This is just to assure the basic idea of the code changes and whether this can theoretically work like this.

Marenz committed:
Add setimmutable and loadimmutable to dialect.

chriseth commented on libyul/backends/evm/EVMAssembly.cpp:
```diff
@@ -215,6 +215,16 @@ AbstractAssembly::SubID EVMAssembly::appendData(bytes const&)
+
+void EVMAssembly::appendImmutableAssignment(std::string const&)
+{
+	yulAssert(false, "LoadImmutable not implemented.");
```
```suggestion
	yulAssert(false, "SetImmutable not implemented.");
```

chriseth commented on libyul/backends/evm/EVMDialect.cpp:
```diff
@@ -170,6 +170,45 @@ map<YulString, BuiltinFunctionForEVM> createBuiltins(langutil::EVMVersion _evmVe
+		builtins.emplace(createFunction(
+			"setimmutable",
+			2,
+			1,
```
```suggestion
			0,
```

chriseth commented on libyul/backends/evm/EVMDialect.cpp:
```diff
@@ -170,6 +170,45 @@ map<YulString, BuiltinFunctionForEVM> createBuiltins(langutil::EVMVersion _evmVe
+			"setimmutable",
+			2,
+			1,
+			SideEffects{},
```
```suggestion
			SideEffects{false, false, false, false, true},
```

chriseth commented on libyul/backends/evm/EVMDialect.cpp:
```diff
@@ -170,6 +170,45 @@ map<YulString, BuiltinFunctionForEVM> createBuiltins(langutil::EVMVersion _evmVe
+			) {
+				solAssert(_call.arguments.size() == 2, "");
+
+				_visitArguments();
```
Hm, we should actually only visit the second argument, right?

Marenz replied:
Yeah, I thought so too. Maybe I need to modify the parameter to allow specifying which one to visit

chriseth commented on libyul/backends/evm/EVMDialect.cpp:
```diff
@@ -170,6 +170,45 @@ map<YulString, BuiltinFunctionForEVM> createBuiltins(langutil::EVMVersion _evmVe
+
+				_visitArguments();
+				Expression const& arg = _call.arguments.front();
+				YulString identifier = std::get<Literal>(arg).value;
```
Do you need the local variable?

chriseth commented on libyul/backends/wasm/WasmDialect.cpp:
```diff
@@ -104,6 +104,8 @@ WasmDialect::WasmDialect()
 
 	addFunction("datasize", {i64}, {i64}, true, {true});
 	addFunction("dataoffset", {i64}, {i64}, true, {true});
+	addFunction("setimmutable", {i32, i32}, {i64}, true, {true, false});
```
Maybe better to leave this for now.

Marenz commented:
Changed code to use a helper function for the `visitArguments` part.

Added the commit of @chriseth 's PR  #8583 

Fixed some cmd line tests that had only whitespace changes.

When running immutable tests with `compileViaYul: also`, we get errors like this:

```
semanticTests/immutable/getter.sol: Exception during test: ../libsolidity/codegen/ir/IRGenerationContext.cpp(86): Throw in function void solidity::frontend::IRGenerationContext::registerImmutableVariable(const solidity::frontend::VariableDeclaration&)
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
std::exception::what: Reserved memory has already been reset.
[solidity::util::tag_comment*] = Reserved memory has already been reset.
```

Will of course squash once approved.

chriseth commented:
Added this commit to https://github.com/ethereum/solidity/pull/8583
